File: blc.pl
$bytemode = !pop; # no command line arguments

sub bit2lam {
  my $bit = pop;
  sub {
    my $x0 = pop;
      sub {
        my $x1 = pop;
        $bit ? $x1 : $x0
      }
  }
}

sub byte2lam {
  my ($bits,$n) = @_;
  sub {
    $n-- ? pop->(sub { bit2lam(vec$bits,$n,1)->(pop) })
              ->(sub { byte2lam($bits,$n)->(pop) }) # cons bitn bits>n
         : sub { pop }                                         # nil
  }
}

sub input {
  my $n = pop;                                # input from n'th character onward
  sub {
    my $c;
    ($B[$n] ||= defined($c=getc) ?
       sub { pop->(sub { ($bytemode ? byte2lam($c,8) : bit2lam($c))->(pop) })
                ->(sub { input($n+1)->(pop) }) # cons charn chars>n
       } :
       sub { sub { pop } }                                           # nil
    )->(pop)
  }
}

sub lam2bit {
  pop->(sub{0})->(sub{1})->()              # force suspension
}

sub lam2byte {
  my $x = pop;	                           # 2nd argument is partial byte
  pop->(sub {
          $x = 2*$x + lam2bit(pop);
          sub {
            my $tail = pop;
            sub { lam2byte($tail, $x) }
          }
        })->(chr $x)                       # end of byte
}

sub output {
  pop->(sub {                              # more chars
          my $c = pop;
          print($bytemode ? lam2byte($c,0) : lam2bit($c));
          sub {
            my $tail = pop;
            sub { output($tail) }
          }
        })->(0)                         # end of output
}

sub getbit {
  $n ||= ($c = getc, $bytemode?8:1);
  vec $c,--$n,1
}

sub program {
  if (getbit()) {             # variable
    my $i;
    $i++ while getbit();
    sub { $_[$i] }
  } elsif (getbit()) {        # application
    my $p=program();
    my $q=program();
    sub {
      my @env = @_;
      $p->(@env)->(sub { $q->(@env)->(pop) }) # suspend argument
    }
  } else {
    my $p = program();
    sub {
      my @env = @_;
      sub { $p->(pop,@env) }  # extend environment with one more argument
    }
  }
}

$| = 1;                                      # non zero value sets autoflush
output program()->()->(input(0))             # run program with empty env on input

File: blc.js
let bytemode = process.argv.length <= 2;
var data;
var nchar = 0;
var nbit = 0;
var progchar;

function bit2lam(bit) {
  return function(x0) {
    return function(x1) {
        return bit ? x1 : x0
      }
  }
}

function byte2lam(bits,n) {
  return function(z) {
    return n-- ? z (function(a) { return bit2lam((bits>>n)&1)(a) })
                   (function(b) { return byte2lam(bits,n)(b) }) // cons bitn bits>n
               : function(y) { return y }                                         // nil
  }
}

function input(n) {           // input from n'th character onward
  return function(z) {
    if (n >= data.length) {
      return function(y) { return y }     // nil
    }
    let c = data[n]; // cons charn chars>n
    return z (function(a) { return (bytemode ? byte2lam(c,8) : bit2lam(c&1))(a) })
             (function(b) { return input(n+1)(b) })
  }
}

function lam2bit(lambit) {
  return lambit(function(){return 0})(function(){return 1})()  // force suspension
  // return lambit(0)(1);
}

function lam2byte(lambits, x) {
  return lambits(function(lambit) {
           y = 2*x + lam2bit(lambit);
           return function(lamtail) {
             return function() { return lam2byte(lamtail, y) }
           }
         })(String.fromCharCode(x))              // end of byte
}

function output(prog) {
  return prog(function(c) {      // more chars
    process.stdout.write(bytemode ? lam2byte(c,0) : lam2bit(c) ? '1' : '0');
    return function(tail) {
      return function() { return output(tail) }
    }
  })(42)                         // end of output
}

function getbit() {
  if (nbit==0) {
    progchar = data[nchar++];
    nbit = bytemode ? 8 : 1;
  }
  let bit = (progchar >> --nbit) & 1;
  return bit;
}

function program() {
  if (getbit()) {               // variable
    var i = 0;
    while (getbit()==1) { i++ }
    return function() { return arguments[i] }
  } else if (getbit()) {        // application
    let p = program();
    let q = program();
    return function(...args) {
      return p(...args)(function(arg) { return q(...args)(arg) }) // suspend argument
    }
  } else {
    let p = program();
    return function(...args) {
      return function(arg) { return p(arg, ...args) }  // extend environment with one more argument
    }
  }
}

process.stdin.on('readable', () => {
  if ((data = process.stdin.read()) != null) {
    prog = program()();
    output(prog(input(nchar)))             // run program with empty env on input
  }
});
